Return actual queue attributes with result from sqs_queue creation/update

emusgrave opened the pull request:
Previously sqs_queue was only returning the requested queue attributes, and not even returning the QueueARN for use elsewhere.  Now it will return the actual attributes in 'results.attributes', which is retrieved with boto's get_queue_attributes() after the queue has been created or updated.

gregdek commented:
Thanks @emusgrave-tricerat. @loia please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

robynbergeron commented:
@loia This change is still pending your review; do you have time to take a look and comment? Please comment with text 'shipit' or 'needs_revision' as appropriate.

wyaeld commented:
@gregdek have a slight concern that @loia only has a single issue comment ever on github, and no commits in last 6 months, so would seem to not be actively maintaining this.  With other useful PRs queuing up on the module, is it desirable to have someone else look at it?

gregdek commented:
@loia still waiting on your review.  Please comment with text 'shipit' or 'needs_revision' as appropriate. If we don't hear from you within 14 days, we will start to look for additional maintainers for this module.

loia commented:
needs_revision

@emusgrave-tricerat Apologies for the delay on this review.

I don't think we should be exposing the raw underlying AWS queue attributes returned by boto for a few reasons:
- The attribute names don't match the module's parameter names which would cause confusion e.g. `delivery_delay` (module) vs `DelaySeconds` (AWS / boto)
- This would be a breaking change to the module's return structure because the attribute names have changed as per above and you've also nested them under a new `attributes` object.
- boto returns additional dynamic state attributes like `ApproximateNumberOfMessages` which don't really make sense for this module.  I think they would be more suitable for a `sqs_queue_facts` module if one existed.

I think it would be better if you simply added the additional attributes you wanted (using the same naming convention) to the return structure and keep it backwards compatible e.g.

```
    if not module.check_mode:
        result['queue_arn'] = queue.get_attributes('QueueArn')['QueueArn']
```

Can you please also add the `return` structure to the documentation?

gregdek commented:
Thanks @emusgrave-tricerat for this PR. A maintainer of this module has asked for revisions to this PR. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

gregdek commented:
@emusgrave-tricerat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

emusgrave commented:
ready_for_review

@loia Thanks for your feedback. I finally got the time to update the PR. 

I have added documentation of the return values and also changed the structure with which they are returned. 
I removed the "attributes" structure and instead directly added the return value of `queue_arn`. In order to return the other values I needed, such as `maximum_message_size`, I had to change the strategy of returning what was passed in as the value of the parameter and instead return the real configuration as returned by AWS/boto. This could be considered a breaking change if someone is relying on the passed in parameter value to be returned, but I think the change more accurately reflects what people will expect to have returned when they make a call to update or create a queue.

gregdek commented:
Thanks @emusgrave-tricerat. To the current maintainers, @loia please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

loia commented:
needs_revision

@emusgrave-tricerat Looks good.  I agree that returning the real AWS configuration is a better behaviour.

Just one minor comment - there's a bit of duplication with populating the attributes for create vs update which I think could be avoided if you move it after the `if/else` block e.g.

```
if queue:
    # Update existing
    ...
else:
    # Create new
    ...

if not module.check_mode:
    result['queue_arn'] = queue.get_attributes('QueueArn')['QueueArn']
    result['default_visibility_timeout'] = queue.get_attributes('VisibilityTimeout')['VisibilityTimeout']
    result['message_retention_period'] = queue.get_attributes('MessageRetentionPeriod')['MessageRetentionPeriod']
    result['maximum_message_size'] = queue.get_attributes('MaximumMessageSize')['MaximumMessageSize']
    result['delivery_delay'] = queue.get_attributes('DelaySeconds')['DelaySeconds']
    result['receive_message_wait_time'] = queue.get_attributes('ReceiveMessageWaitTimeSeconds')['ReceiveMessageWaitTimeSeconds']
```

gregdek commented:
Thanks @emusgrave-tricerat for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@emusgrave A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@emusgrave Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

emusgrave commented:
ready_for_review

gregdek commented:
Thanks @emusgrave. To the current maintainers, @loia please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

epa095 commented:
It looks like @loia  has not really been active for some time. Is there any way to get this accepted? Ansible-SQS is a lot less useful now when we dont get the ARN.

loia commented:
Sorry for the delay guys..

shipit

gregdek commented:
Thanks again to @emusgrave for this PR, and thanks @loia for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented on cloud/amazon/sqs_queue.py:
```diff
@@ -113,16 +148,23 @@ def create_or_update_sqs_queue(connection, module):
             result['changed'] = True
-
+        
+        if not module.check_mode:
```
AFAIKS wouldn't it make sense to check for queue instad of module.check_mode? I mean queue maybe exists and so the results can be done as well. right? Did I miss anything? 

``` diff
-        if not module.check_mode:
+        if queue:
```

loia replied:
Yes, that would technically work.

It may look a bit strange though that the attributes have their original values instead of the updated ones.  To be accurate, you may need to load the original values, then override with the request values.

I'll leave it up to @emusgrave if he wants to include this or whether it should be a separate PR.

CuriousLearner commented on cloud/amazon/sqs_queue.py:
```diff
@@ -113,16 +148,23 @@ def create_or_update_sqs_queue(connection, module):
             result['changed'] = True
-
+        
+        if not module.check_mode:
```
Hi, when I'm creating queue, although the queue is successfully created, I don't get the `queue_arn` from ansible. Can I force this attribute to always return `queue_arn`?
